okio/src/main/java/okio/Buffer.java

```diff
@@ -130,6 +139,26 @@ public long size() {
     };
   }
 
+  @Override public Reader reader() {
+    return new BomAwareReader(
+        null, BOM_UTF_8, BOM_UTF_32_LE, BOM_UTF_32_BE, BOM_UTF_16_LE, BOM_UTF_16_BE);
+  }
+
+  @Override public Reader reader(Charset charset) {
+    if (charset == null) throw new IllegalArgumentException("charset == null");
+
+    String charsetName = charset.name();
+    if (charsetName.equals("UTF-8")) {
+      return new BomAwareReader(charset, BOM_UTF_8);
+    } else if (charsetName.startsWith("UTF-16")) {
+      return new BomAwareReader(charset, BOM_UTF_16_BE, BOM_UTF_16_LE);
```

Comment:
Hah. I just assumed InputStreamReader has no BOM handling at all. But it looks like at least on JRE 1.8 it works just fine for UTF-16/UTF-32 (BOM is used to detect endianness and is not part of the output). The only issue is UTF-8 + BOM: http://bugs.java.com/view_bug.do?bug_id=4508058

This might make things considerably easier. I'll investigate further.